### PR TITLE
Sliver padding overlap fix

### PR DESCRIPTION
**Pull request by Kavantix**
## Description

This PR fixes overlap getting broken by SliverPadding

## Related Issues

Improves #28158

## Tests

I added the following tests:

A test to ensure that the overlap passed to the child sliver is reduced by the amount of padding instead of set to 0

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Kavantix** [comment]
The comment on the top of the SliverPadding should be updated.
Is there any other example of SliverPadding not working correctly (since the pinned header now should work as expected)

**Comment by Piinks** [comment]
Hi @Kavantix thank you for the contribution! 

> The comment on the top of the SliverPadding should be updated.

Please update the documentation you mentioned as part of this change.

> Is there any other example of SliverPadding not working correctly (since the pinned header now should work as expected)

Since this should fix the pinned header, can you also add a widget test that verifies that specific use case? That will also ensure we don't regress later. :)

**Comment by Piinks** [comment]
Might be related to https://github.com/flutter/flutter/issues/55170

**Comment by Kavantix** [comment]
@Piinks I asked about other examples since I think there must be others but atm I’m not aware of any
So should I just change the comment by removing the example?

Also 55170 is not related, I’ve looked at the underlying code and I’ll post my findings there

**Comment by Kavantix** [comment]
I have been testing some more with cases of the padding.
I can manage to improve it somewhat by making the overlap even more complex thanks to overlap being negative when overscrolling on iOS. And adding paintorigin, but it still has some quirks I cannot get out.
All of the issues are only in the main extent though

**Comment by Kavantix** [comment]
@googlebot I fixed it.

**Comment by Kavantix** [comment]
Just to clarify,

Because of these remaining issue I think it's not really helpful to add a widget test for the persistent header case until we can make sure it works exactly as wanted, especially since I currently see no nice option to resolve them.

I did update the documentation comment and add tests for the other changes (negative overlap and paintorigin)

**Comment by Piinks** [comment]
> Because of these remaining issue I think it's not really helpful to add a widget test for the persistent header case until we can make sure it works exactly as wanted, especially since I currently see no nice option to resolve them.

That sounds like this doesn't actually solve the issue then?

**Comment by Kavantix** [comment]
Not completely no.
But it does improve it quite a bit.
And more importantly makes SliverPadding handle all the Sliver properties correctly.
The issues that remain are inherent of the current SliverGeometry/SliverConstraints behaviour since the child would somehow need to communicate to sliverpadding that it should stay pinned for instance, however that seems out of scope for this PR.
When adding padding solely in the cross axis this PR does fix the issue.

File: packages/flutter/test/widgets/slivers_padding_test.dart [hunk 1]
```diff
@@ -459,4 +471,79 @@ void main() {
...
   });
+
+  testWidgets('SliverPadding consumes only its padding from the overlap of its parent\'s constraints', (WidgetTester tester) async {
+    final _MockRenderSliver mock = _MockRenderSliver();
+    final RenderSliverPadding renderObject = RenderSliverPadding(
+      padding: const EdgeInsets.only(top: 20),
+    );
+    renderObject.child = mock;
+    renderObject.layout(const SliverConstraints(
+         viewportMainAxisExtent: 100.0,
```
Comment (Piinks) [inline packages/flutter/test/widgets/slivers_padding_test.dart:482]: Style nits: The indentations on all of these SliverConstraints is off.

**Review by Piinks — APPROVED** [review]
LGTM with nits. :)
Thanks for the contribution! It looks like the web tests passed, but did not report the right status. I've re-run them to see if that should clear it up.
I am curious to see what the Google testing yields before we land this, it's triggered by LGTM. 🎉

File: packages/flutter/lib/src/rendering/sliver_padding.dart [hunk 2]
```diff
@@ -19,11 +19,9 @@ import 'sliver.dart';
 /// passed on to the child.
 ///
 /// {@template flutter.rendering.sliverPadding.limitation}
-/// Applying padding to anything but the most mundane sliver is likely to have
+/// Applying padding in the main extent of the viewport to anything but the most mundane sliver is likely to have
 /// undesired effects. For example, wrapping a [RenderSliverPinnedPersistentHeader]
-/// will cause the app bar to overlap earlier slivers (contrary to the normal
-/// behavior of pinned app bars), and while the app bar is pinned, the padding
-/// will scroll away.
+/// will cause the app bar to only the appbar to stay pinned while the padding will scroll away.
```
[the remark this answers is not in the document]
Comment (Kavantix) [reply packages/flutter/lib/src/rendering/sliver_padding.dart]: This was not a nit at all, thanks for pointing it out :)

**Comment by Kavantix** [comment]
Fixed the constraints code layout and also fixed the duplicated comment on sliverpadding

**Comment by Piinks** [comment]
Looks like the customer tests are failing for an unrelated reason, can you update your branch with the latest from master?

**Comment by Kavantix** [comment]
Merged master into my branch

**Comment by Piinks** [comment]
Thanks @Kavantix for updating again, I am not sure why the docs shard is still failing, it appears unrelated to this change. @goderbauer do you recognize this docs error?

**Comment by goderbauer** [comment]
The error looks like a flake to me. I re-ran the check. Hopefully it will pass now.

**Comment by Kavantix** [comment]
Seems like it passed now🎉

**Pull request merged**